Allow passing `ParseOptions` to inline tests

ntBre opened the pull request:
## Summary

The changes here are based on the similar behavior in biome's [`collect_tests`](https://github.com/biomejs/biome/blob/b9f8ffea9967b098ec4c8bf74fa96826a879f043/xtask/codegen/src/parser_tests.rs#L159) function, which allows a syntax for inline test headers like

```
label language name [options]
```

Before this PR, we only allowed headers of the form `label name`, where `label` is either `test_err` or `test_ok` and `name` is the name of the test. This PR adds support for an optional, trailing `options` field, corresponding to JSON-serialized `ParseOptions`. These get written to a `*.options.json` file alongside the inline test script and read when that test is run.

This is currently stacked on #16090 so that I had something to test.

I wanted to add `serde` as a dev dependency only, but since these are run as integration tests, I had to add it as a full dependency. For this exact use case, we could just parse a trailing version string like "3.10" but I thought we might want the full `ParseOptions` for the future.

## Test Plan

Existing inline tests, plus two new inline tests for `match-before-py310`.

MichaReiser commented on crates/ruff_python_parser/src/parser/statement.rs:
```diff
@@ -2265,6 +2265,16 @@ impl<'src> Parser<'src> {
 
         let range = self.node_range(start);
 
+        // test_err match_before_py310 { "target_version": "3.9" }
+        // match 2:
+        //     case 1:
+        //         pass
```
I find the options json files somewhat noisy. I know, we do the same for the formatter and maybe that was already a bad choice from my side. 

Should we use a pragma comment instead?

```python
# parserOptions: { target_version: "3.10" }
```

and they are only allowed on the first line of the test.

dhruvmanila replied:
Are you saying that the comment would be part of the source code? Like the following:
```py
# parse_options: { target_version: "3.10" }
match 2:
	case 1:
		pass
```

That is a good alternative but then I would recommend (not in this PR) to add the file content (if it's not too noisy) in the snapshot to make sure that the snapshot contains all the necessary details (content, parse options) to validate it.

MichaReiser commented on crates/ruff_python_parser/tests/snapshots/invalid_syntax@match_before_py310.py.snap:
```diff
@@ -0,0 +1,64 @@
+1 | / match 2:
+2 | |     case 1:
+3 | |         pass
+  | |____________^ Syntax Error: Cannot use `match` statement on Python 3.9 (syntax was new in Python 3.10)
```
```suggestion
  | |____________^ Syntax Error: Cannot use `match` statement on Python 3.9 (syntax was added in Python 3.10)
```

ntBre replied:
I'm going to apply this one to #16090.

MichaReiser commented on crates/ruff_python_parser/Cargo.toml:
```diff
@@ -22,6 +22,7 @@ bstr = { workspace = true }
 compact_str = { workspace = true }
 memchr = { workspace = true }
 rustc-hash = { workspace = true }
+serde = { workspace = true }
```
We should make `serde` an optional dependency. Not all downstream users need serde serialization

MichaReiser commented on crates/ruff_python_parser/Cargo.toml:
```diff
@@ -13,7 +13,7 @@ license = { workspace = true }
 
 [dependencies]
-ruff_python_ast = { workspace = true }
+ruff_python_ast = { workspace = true, features = ["serde"] }
```
We should only enable the `serde` feature for dev builds

MichaReiser reviewed (commented):
This looks good. I do think I'd prefer a pragma comment over an external `options.json` file

dhruvmanila reviewed (commented):
This is great! Thanks for taking this on quickly and implementing it.

dhruvmanila commented on crates/ruff_python_parser/tests/fixtures.rs:
```diff
@@ -77,11 +82,20 @@ fn test_valid_syntax(input_path: &Path) {
 /// Snapshots the AST and the error messages.
 fn test_invalid_syntax(input_path: &Path) {
+    let options_path = input_path.with_extension("options.json");
+    let options = fs::read_to_string(options_path)
+        .ok()
+        .and_then(|s| serde_json::from_str(&s).ok())
+        .unwrap_or_else(|| {
+            ParseOptions::from(Mode::Module).with_target_version(PythonVersion::PY313)
+        });
```
This won't be required if we go with the pragma comment instead of `.options.json`

ntBre commented:
I rebased on #16090 to get the new `SyntaxError` name and switched to the pragma implementation. It seems quite nice and ended up being less code too. The only piece I might be missing is including the code file contents in the snapshot, but I wasn't quite sure if that was still needed. It sounds like that could be a follow-up either way.

I made `serde` an optional dependency gated behind the `serde` feature in the parser. You have to pass `--features serde` to run the parser tests with something like `cargo test -p ruff_python_parser --features serde`, but I think CI runs everything with `--all-features`. This feels a bit hacky but maybe it's okay? I left that commit last so it would be easy to revert.

ntBre commented:
My other branch must be missing Doug's alist changes, at least I assume that's where the codspeed failure is coming from. I expect that to go away when I change the base branch to `main`.

ntBre commented:
While trying to use this branch as a base for detecting some syntax errors, I realized that I forgot to actually check for the new errors in the *valid* syntax case. It's difficult to test this case because it requires a version-related syntax error in a `test_ok` inline test. I guess a `#[should_panic]` test calling `test_valid_syntax` directly could work, but I can also confirm that these changes now detect the false-positive I was debugging on the other branch.

MichaReiser commented on crates/ruff_python_parser/src/lib.rs:
```diff
@@ -661,6 +661,11 @@ impl FusedIterator for TokenIterWithContext<'_> {}
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Deserialize),
+    serde(rename_all = "lowercase")
```
We normally use kebab case everywhere

```suggestion
    serde(rename_all = "kebab_case")
```

dhruvmanila replied:
I think it's `kebab-case` (hyphen, not an underscore) (https://serde.rs/container-attrs.html#rename_all)

MichaReiser commented on crates/ruff_server/src/lint.rs:
```diff
@@ -199,6 +199,22 @@ pub(crate) fn check(
             .flatten(),
     );
 
+    let lsp_diagnostics = lsp_diagnostics.chain(
+        show_syntax_errors
+            .then(|| {
+                parsed.unsupported_syntax_errors().iter().map(|error| {
+                    unsupported_syntax_error_to_lsp_diagnostic(
+                        error,
+                        &source_kind,
+                        locator.to_index(),
+                        encoding,
+                    )
+                })
+            })
+            .into_iter()
+            .flatten(),
+    );
```
I don't understand this change. Why are we appending the `unsupported_syntax_errors` on line 189 and here?

dhruvmanila replied:
I think this PR probably just needs to be rebased on `main`?

ntBre replied:
Oh yes, I was using git too late last night!

dhruvmanila commented on crates/ruff_python_parser/src/parser/options.rs:
````diff
@@ -21,10 +21,13 @@ use crate::{AsMode, Mode};
 /// let options = ParseOptions::from(PySourceType::Python);
 /// ```
 #[derive(Clone, Debug)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
````
I'd use `#[serde(rename_all = "kebab-case")]` for consistency

ntBre commented:
Based on a discussion with @MichaReiser on Discord, I backed out the `serde` feature changes and added `JsonParseOptions` and `JsonMode` types in `fixtures.rs` to keep `serde` as a `dev-dependency`.

This avoids needing to run the parser integration tests with the `serde` feature enabled or requiring `serde` as a full dependency.
